selfhost/typechecker.jakt

```diff
@@ -1729,18 +1729,34 @@ struct Typechecker {
         .current_function_id = method_id
 
         let VOID_TYPE_ID = builtin(BuiltinType::Void)
+        mut function_return_type_id = .typecheck_typename(
+            parsed_type: func.return_type,
+            scope_id: function_scope_id,
+            name: None
+        )
 
-        let block = .typecheck_block(parsed_block: func.block, parent_scope_id: function_scope_id, safety_mode: SafetyMode::Safe)
-        let function_return_type_id = .typecheck_typename(parsed_type: func.return_type, scope_id: function_scope_id, name: None)
+        if not func.is_fat_arrow and func.return_type is Empty {
+             function_return_type_id = VOID_TYPE_ID
+        }
+
+        checked_function.return_type_id = function_return_type_id
+
+        let block = .typecheck_block(
+            parsed_block: func.block,
+            parent_scope_id: function_scope_id,
+            safety_mode: SafetyMode::Safe
+        )
 
         mut return_type_id = function_return_type_id
         if function_return_type_id.equals(unknown_type_id()) and not block.statements.is_empty() {
-            // If the return type is unknown, and the function starts with a return statement,
+            // If the return type is unknown, function is a fat arrow function and it starts with a return statement,
```

Comment:
I'm a little confused with this statement. If a function is a fat arrow, then the thing it's pointing to should be the value it's returned.

Is this talking about the case where there are additional returns inside of the fat arrow?

Reply:
I've changed the comment, added a copy of it to the typecheck_function and unified the logic of that snippet between the typecheck_method and typecheck_function

Reply:
Also: unified this logic for functions with generic parameters